datalad_next/utils/deprecate.py

```diff
@@ -0,0 +1,89 @@
+import inspect
+import warnings
+from functools import wraps
+
+from datalad_next.utils import get_wrapped_class
+
+
+def deprecated(msg, version, parameter=None, parameter_choice=None):
+    """Annotate functions, classes, or function parameters with standardized
+    deprecation warnings.
+
+    Parameters
+    ----------
+    version: str
+      Software version number at which the deprecation was made
+    msg: str
+      Custom message to append to a deprecation warning
+    parameter: str
+      Individual parameter being deprecated (instead of entire function/class)
+    parameter_choice: str
+      Individual parameter choice of specified parameter being deprecated
+    """
+
+    base_template = "{func} was deprecated in version {version}. {msg}"
+    if parameter is None:
+        # the entire class/function is deprecated
+        def decorator(func):
+            @wraps(func)
+            def func_with_deprecation_warning(*args, **kwargs):
+                fname = (
```

Comment:
Ahhh, no. I am sorry. The renaming only applies to the `Dataset` method.

Hence, two things instead:
- include a check, that the class is a `subclass` of `Interface` to make sure what is decorated is a command implementation, so that the logic actually applies
- ideally transform the class name to the command name. That transformation should already exist in `interface`. (I would, however, not consider this point a blocker)

Reply:
Alternative, since this is making it more complex:

Rip out that logic. When a command is to be deprecated entirely, currently the class needs to be decorated and done. If someone decorates the wrong thing - not the problem of this decorator, I guess.

Reply:
I keep thinking about this. The issue with the simplest (the last) suggestion is, that a class deprecation would be triggered by importing the class, not by calling it. I wonder, whether that's actually a good thing. It could lead to the next step of forcing us to make imports leaner. If the deprecator would register that class in a singleton - say `datalad._depracted`, we could probably implement a delayed import into api. Just like extensions aren't necessarily imported on every datalad start up.

Reply:
Thanks for your thoughts, although I'm not 100% certain I can follow them :)
Just for completeness, here is the current state:

> The parameter value overwrite-existing of parameter force of <class 'datalad_next.commands.download.Download'> was deprecated in version 1.0.0.

An easy alternative could also be to leave any messaging entirely to the user. So if you want to deprecate one parameter option, the ``msg`` could become "The option ABC of parameter D is deprecated. Use EFG instead". And then we wouldn't do any templating, but simply pass down the message into the deprecation warning

Reply:
I have difficulties grasping the issue here. I think it would make sense to actually test the behavior that is discussed here and make it obvious, and verified. Afterwards it would be easier to debate whether it is problematic.